src/networking.c

```diff
@@ -1591,90 +1591,71 @@ int _writeToClient(client *c, ssize_t *nwritten) {
-                    break;
-                }
-                remaining -= (ssize_t)(o->used - c->sentlen);
+        if (c->bufpos > 0) ++nbuf; /* include the static reply buffer. */
+        struct iovec iov[nbuf > IOV_MAX ? IOV_MAX : nbuf];
+        int iovcnt = 0;
+        size_t cum_bytes = 0;
+        /* If the static reply buffer is not empty, 
+         * add it to the buffer queue of writev() as well. */
+        if (c->bufpos > 0) {
+            iov[iovcnt].iov_base = c->buf + c->sentlen;
+            iov[iovcnt].iov_len = c->bufpos - c->sentlen;
+            cum_bytes += iov[iovcnt++].iov_len;
+        }
+        /* The first node of reply list might be incomplete from the last call,
+         * thus it needs to be calibrated to get the actual data address and length. */
+        size_t offset = c->bufpos > 0 ? 0 : c->sentlen;
+        listIter iter;
+        listNode *next;
+        clientReplyBlock *o;
+        listRewind(c->reply, &iter);
+        while ((next = listNext(&iter)) && iovcnt < IOV_MAX && cum_bytes < NET_MAX_WRITES_PER_EVENT) {
+            o = listNodeValue(next);
+            if (o->used == 0) { /* empty node, just release it and skip. */
                 c->reply_bytes -= o->size;
                 listDelNode(c->reply, next);
-                c->sentlen = 0;
+                offset = 0;
+                continue;
             }
-        } else {
-            clientReplyBlock *o = listNodeValue(listFirst(c->reply));
-            size_t objlen = o->used;
 
-            if (objlen == 0) {
-                c->reply_bytes -= o->size;
-                listDelNode(c->reply,listFirst(c->reply));
-                return C_OK;
-            }
+            iov[iovcnt].iov_base = o->buf + offset;
+            iov[iovcnt].iov_len = o->used - offset;
+            cum_bytes += iov[iovcnt++].iov_len;
+            offset = 0;
+        }
+        if (iovcnt == 0) return C_OK;
+        *nwritten = connWritev(c->conn, iov, iovcnt);
+        if (*nwritten <= 0) return C_ERR;
 
-            *nwritten = connWrite(c->conn, o->buf + c->sentlen, objlen - c->sentlen);
-            if (*nwritten <= 0) return C_ERR;
-            c->sentlen += *nwritten;
+        /* Locate the new node which has leftover data and
+         * release all nodes in front of it. */
+        ssize_t remaining = *nwritten;
+        if (c->bufpos > 0) { /* deal with static reply buffer first. */
+            int buf_len = c->bufpos - c->sentlen;
 
-            /* If we fully sent the object on head go to the next one */
-            if (c->sentlen == objlen) {
-                c->reply_bytes -= o->size;
-                listDelNode(c->reply, listFirst(c->reply));
+            /* If the buffer was sent, set bufpos to zero to continue with
+             * the remainder of the reply. */
+            if ((int)(c->sentlen + remaining) >= c->bufpos) {
+                c->bufpos = 0;
                 c->sentlen = 0;
             }
+            remaining -= buf_len;
```

Comment:
aren't we now missing the update of `c->sentlen` in case we didn't send the full thing (an `else` for the above `if`?)

Reply:
Delete it in the latest commit by accident, added back.